great_expectations/data_context/store/datasource_store.py

```diff
@@ -123,6 +125,28 @@ def delete_by_name(self, datasource_name: str) -> None:
         )
         self.remove_key(datasource_key)
 
+    def delete(self, datasource_config: DatasourceConfig) -> None:
+        """Deletes a DatasourceConfig persisted in the store using its config.
+
+        Args:
+            datasource_config: The config of the Datasource to delete.
+        """
+
+        self.remove_key(self._build_key_from_config(datasource_config))
+
+    def _build_key_from_config(
+        self, datasource_config: DatasourceConfig
+    ) -> Union[GeCloudIdentifier, DataContextVariableKey]:
+        if hasattr(datasource_config, "id_"):
+            id_ = datasource_config.id_
+        else:
+            id_ = None
+        if hasattr(datasource_config, "name"):
+            name = datasource_config.name
+        else:
+            name = None
+        return self.store_backend.build_key(name=name, id_=id_)
```

Comment:
I foresee this logic being replicated quite a bit - it may make sense to refactor it once we add more id/name-support objs.